File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """

        def is_valid_sequence(sequence):
            # Filter out the empty cells '.'
            nums = [num for num in sequence if num != '.']
            return len(nums) == len(set(nums))

        # Check each row
        for row in board:
            if not is_valid_sequence(row):
                return False

        # Check each column
        for col in range(9):
            if not is_valid_sequence([board[row][col] for row in range(9)]):
                return False

        # Check each 3x3 box
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                box = [board[x][y] for x in range(i, i+3) for y in range(j, j+3)]
                if not is_valid_sequence(box):
                    return False

        return True


        
        

        

File: 0036-valid-sudoku/0036-valid-sudoku.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut sets: Vec<HashSet<char>> = vec![HashSet::new(); 27];

        for row in 0..9 {
            for col in 0..9 {
                let num = board[row][col];
                    if num != '.' {
                        let row_idx = row;
                        let col_idx = col + 9;
                        let box_idx = 18 + (row / 3) * 3 + (col / 3);

                        if sets[row_idx].contains(&num) || sets[col_idx].contains(&num) || sets[box_idx].contains(&num){
                            return false
                        }

                        sets[row_idx].insert(num);
                        sets[col_idx].insert(num);
                        sets[box_idx].insert(num);
                    }
            }
        }
        true

        
    }
}